### PR TITLE
docs: define vime integration design

### LINKED ISSUE #119: [WS5] RL-Kernel to vime integration design doc
Part of #83; depends on #118; feeds the P1 full-integration effort

Produce the "RL-Kernel to vime integration" design doc. The goal is to define a non-intrusive integration path and positioning statement before any broad implementation work starts.

Planned:
- [ ] Integration design doc covering insertion points, expected tensor/API contracts, lifecycle, ownership boundaries, and rollout/training-side responsibilities.
- [ ] Non-intrusive activation path: custom op / optional backend / config flag shape, with pure vime inference and native RL paths unaffected when disabled.
- [ ] Change-surface estimate: what can live entirely in RL-Kernel, what requires vime/slime adapter code, and what would need upstream vLLM/vllm-router changes.
- [ ] Positioning statement: RL-Kernel provides operator-level consistency beneath vime's framework-level alignment; complementary rather than competing.
- [ ] Risk and fallback section: private API reliance, version pinning, missing extension points, unsupported hardware, and how the PoC should fail safely.

Non-goals: no production integration, no vime upstream PR, and no claim that this closes real-engine train-inference consistency. WS5 scopes the path; WS3/P1 own the full real-engine integration.

Start after the hook-point inventory in #118 is usable.

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary

- Add the WS5 RL-Kernel to Vime integration design doc for #119.
- Define the non-intrusive activation model, ownership boundaries, lifecycle, tensor/API contracts, and rollout/training-side responsibilities.
- Capture the change-surface estimate across RL-Kernel, Vime/slime adapter work, and upstream vLLM/vllm-router requirements.
- Document risk/fallback behavior and the recommended #120 PoC path.

## Scope

This is a documentation-only PR.

It does not add a production integration, change Vime/slime/vLLM/Megatron source, inject RL-Kernel operators, or claim real-engine train-inference consistency.

Depends on #118  
Closes #119  
Part of #83

### CODE REVIEW

File: docs/design/vime-integration-design.md [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+| `token_ids` | `[N]` or same leading shape as `logits` | int64/int32 accepted by wrapper | Target token ids aligned to response logits. |
+| output | `[N]` or original leading shape | fp32 preferred for comparison | One selected logprob per response token. |
+
+The adapter must handle Vime's layouts before calling RL-Kernel:
+
+- THD and BSHD packing;
+- context parallel slicing;
+- tensor parallel vocabulary partitioning;
+- optional all-gather-CP mode;
+- temperature scaling with `args.rollout_temperature`;
```
Comment (z1ying) [inline docs/design/vime-integration-design.md:209]: I think this could be replaced with:

- **Temperature scaling**: the adapter divides logits by `args.rollout_temperature` before invoking the op. The RL-Kernel logp op is temperature-unaware and must not apply scaling itself; otherwise the parity comparison against Vime's native `compute_log_probs` would be made on differently-scaled logits.

File: docs/design/vime-integration-design.md [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+Fallback rules:
+
+1. `off` means Vime runs exactly as it does today.
+2. `observe` may import RL-Kernel and emit diagnostics, but must not change
+   tensors consumed by training.
+3. `train_ops` may replace training-side operator calls only after tensor
+   contracts pass validation.
+4. `rollout_ops` is blocked for selected-logprob recomputation until vLLM or
+   vllm-router exposes logits or an internal custom-op insertion point.
+5. In non-strict mode, any import, shape, dtype, device, or backend failure logs
```
Comment (z1ying) [inline docs/design/vime-integration-design.md:99]: Who implements this? `get_op` raises `RuntimeError` (registry.py:315) and shape errors raise at call time (pytorch/loss/logp.py:28) — nothing in RL-Kernel catches them, so shouldn't the doc say the adapter owns this try/except, not "RL-Kernel owns local fallback"?

File: docs/design/vime-integration-design.md [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+An adapter should lazily import RL-Kernel at the first enabled callsite:
+
+```python
+try:
+    from rl_engine.kernels.registry import kernel_registry
+except Exception as exc:
+    fallback_or_raise(exc)
+```
+
+Do not import or initialize CUDA extensions on Vime's driver process unless the
```
Comment (z1ying) [inline docs/design/vime-integration-design.md:274]: Doesn't this import already violate the constraint below? `registry.py:337` instantiates `KernelRegistry()` at import time, which calls `torch.cuda.get_device_capability()` (registry.py:255) → `_lazy_init()` → CUDA context on the driver process.

File: docs/design/vime-integration-design.md [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    current_logps,
+    old_logps,
+    ref_logps,
+    rewards,
+    completion_mask,
+    clip_eps=...,
+    beta=...,
+    samples_per_prompt=...,
+)
+```
```
Comment (z1ying) [inline docs/design/vime-integration-design.md:233]: This snippet no longer matches main: both `NativeGRPOLossOp` and `TritonGRPOLossOp` now take `(policy_logits, ref_logits, action_ids, old_logps, rewards, completion_mask, ...)` — logits-based via the fused ratio/KL op, not `current_logps/ref_logps`.

Two contract points worth capturing here:
1. `__call__`/`forward` expect *raw* rewards and group-normalize internally, while L238 shows Vime supplies already-normalized `batch["advantages"]` — feeding those as `rewards` double-normalizes. Both ops expose `apply(..., sample_advantages, ...)` for precomputed advantages ("`apply` takes the per-sequence advantage vector directly", ops/triton/loss/grpo_loss.py:65-66); the doc should direct adapters there.
2. The logits-based signature also changes the tensor contract below (`[B, T]` → logits are `[B, T, V]`), which affects the packing-adapter estimate.

File: docs/design/vime-integration-design.md [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+selected_logps = logp_op(logits, token_ids)
+```
+
+Contract for a Vime adapter:
+
+| Tensor | Shape | Dtype | Notes |
+| --- | --- | --- | --- |
+| `logits` | `[N, V]` or reshapeable to that form | fp32/bf16 depending on backend | Must be contiguous for fused CUDA paths. |
+| `token_ids` | `[N]` or same leading shape as `logits` | int64/int32 accepted by wrapper | Target token ids aligned to response logits. |
+| output | `[N]` or original leading shape | fp32 preferred for comparison | One selected logprob per response token. |
```
Comment (z1ying) [inline docs/design/vime-integration-design.md:201]: "fp32 preferred" is worth pinning down: the default `apply`/`__call__` entries return the *input* dtype (e.g. `NativeLogpOp.apply` uses `output_dtype=logits.dtype`); fp32 output requires calling `apply_fp32` explicitly. Should the contract state that parity comparisons use `apply_fp32`? Relatedly, since this section motivates parity/drift testing, the separate `logp_deterministic` op type (batch-invariant, fp32 by design, registry.py:158-160) seems like the natural reference implementation to mention here.

**Review by z1ying — COMMENTED** [review]
Thanks for the PR! I've left a few suggestions for clarification.
